pretty/src/geometry/rect.rs

```diff
@@ -46,6 +73,126 @@ impl Rect {
         }
         v.into_iter()
     }
+
+    /// Get the number of columns in the rectangle
+    pub fn width(&self) -> Col {
+        self.cols.len()
+    }
+
+    /// Get the number of rows in the rectangle
+    pub fn height(&self) -> Row {
+        self.rows.len()
+    }
+
+    /// Get the size of the rectangle
+    pub fn size(&self) -> Pos {
+        Pos {
+            row: self.rows.len(),
+            col: self.cols.len(),
+        }
+    }
+
+    /// Split the rectangle into two sub-rectangles, where the left
+    /// sub-rectangle has `left_width` columns. Returns None if `left_width` is
+    /// greater than the width of the original rectangle.
+    pub fn horz_split(&self, left_width: Col) -> Option<(Rect, Rect)> {
+        let (left_cols, right_cols) = self.cols.split(left_width)?;
+        Some((
+            Rect {
+                cols: left_cols,
+                rows: self.rows,
+            },
+            Rect {
+                cols: right_cols,
+                rows: self.rows,
+            },
+        ))
+    }
+
+    /// Split the rectangle into two sub-rectangles, where the top
+    /// sub-rectangle has `top_height` columns. Returns None if `top_height` is
+    /// greater than the height of the original rectangle.
+    pub fn vert_split(&self, top_height: Row) -> Option<(Rect, Rect)> {
+        let (top_rows, bottom_rows) = self.rows.split(top_height)?;
+        Some((
+            Rect {
+                rows: top_rows,
+                cols: self.cols,
+            },
+            Rect {
+                rows: bottom_rows,
+                cols: self.cols,
+            },
+        ))
+    }
+
+    /// Given N `widths`, returns an iterator over N sub-rectangles with those
+    /// widths, in order from left to right. `next()` will panic if the next
+    /// width is larger than the width of the remaining rectangle.
+    pub fn horz_splits<'a>(&self, widths: &'a [Col]) -> SplitRectHorzIter<'a> {
+        SplitRectHorzIter {
+            remaining: *self,
+            widths,
+        }
+    }
+
+    /// Given N `heights`, returns an iterator over N sub-rectangles with those
+    /// heights, in order from top to bottom. `next()` will panic if the next
+    /// height is greater than the height of the remaining rectangle.
+    pub fn vert_splits<'a>(&self, heights: &'a [Row]) -> SplitRectVertIter<'a> {
+        SplitRectVertIter {
+            remaining: *self,
+            heights,
+        }
+    }
+}
+
+impl<'a> Iterator for SplitRectHorzIter<'a> {
+    type Item = Rect;
+
```

Comment:
Instead of implementing this on `Rect`, implement it on `Range`, and use `.map()` to produce an `impl Iterator` for the two `Rectangle` methods.

Reply:
done